### PR TITLE
Add Python fallback to random_k_out_graph + document dependencies

### PR DESCRIPTION
**Pull request by rossbar**
Intended to address the first bullet of #7716 

Adds a try-except-based fallback to the pure-Python implementation for `random_k_out_graph` when NumPy is not installed. Also adds some documentation on the default behavior, as well as recipes to select the specific implementation for niche use-cases. Also adds a prominent `.. note` near the top of the docstring to highlight this behavior.

### CODE REVIEW

**Review by dschult — APPROVED** [review]
Looks good!  Thanks!!

File: networkx/generators/directed.py [hunk 1]
```diff
@@ -423,6 +423,9 @@ def sample(v, nodes):
 def random_k_out_graph(n, k, alpha, self_loops=True, seed=None):
     """Returns a random `k`-out graph with preferential attachment.
 
+    .. note:: Different implementations will be used based on whether NumPy is
+       available. See Notes for details.
```
Comment (MridulS) [inline networkx/generators/directed.py:427]: Should we put in an explicit version number for this change?

File: networkx/generators/directed.py [hunk 2]
```diff
@@ -423,6 +423,11 @@ def sample(v, nodes):
...
+
+       .. versionchanged:: 3.5
```
Comment (Schefflera-Arboricola) [inline networkx/generators/directed.py:429]:
```suggestion
    .. versionchanged:: 3.5
       Different implementations will be used based on whether NumPy is
       available. See Notes for details.
```

File: networkx/generators/directed.py [hunk 3]
```diff
@@ -478,18 +483,37 @@ def random_k_out_graph(n, k, alpha, self_loops=True, seed=None):
     The returned multidigraph may not be strongly connected, or even
     weakly connected.
 
+    `random_k_out_graph` has two implementations: an array-based formulation that
+    uses `numpy` (``_random_k_out_graph_numpy``), and a pure-Python
+    implementation (``_random_k_out_graph_python``).
+    The NumPy implementation is more performant, especially for large `n`, and is
+    therefore used by default. If NumPy is not installed in the environment,
+    then the Python version is used.
```
Comment (Schefflera-Arboricola) [inline networkx/generators/directed.py:491]:
```suggestion
    then the pure Python implementation is executed.
```

File: networkx/generators/directed.py [hunk 4]
```diff
@@ -478,18 +483,37 @@ def random_k_out_graph(n, k, alpha, self_loops=True, seed=None):
...
+    If you have a use-case which requires explicit control over which implementation
+    is used, call the corresponding function directly::
```
Comment (Schefflera-Arboricola) [inline networkx/generators/directed.py:493]:
```suggestion
    However, you can explicitly control which implementation is executed by directly
    calling the corresponding function::
```

**Review by Schefflera-Arboricola — APPROVED** [review]
LGTM! Thanks @rossbar, @dschult, @MridulS and @ruidazeng !

CI failure fixed in https://github.com/networkx/networkx/pull/7746

**Pull request merged**